src/spec-configuration/containerFeaturesOCI.ts

```diff
@@ -187,34 +208,232 @@ export async function getFeatureBlob(output: Log, env: NodeJS.ProcessEnv, url: s
+    output.write(`Uploading blob: ${name} with digest ${digest}`, LogLevel.Trace);
+
+    // const headers: HEADERS = {
+    //     'user-agent': 'devcontainer',
+    //     'authorization': `Bearer ${registryAuthToken}`
+    // }
+
+    let url = '';
+    if (blobPutLocationUriPath.startsWith('https://')) {
+        url = blobPutLocationUriPath;
+    } else {
+        url = `https://${featureRef.registry}${blobPutLocationUriPath}`;
+    }
+    url += `?digest=sha256:${digest}`;
+
+    output.write(`Crafted blob url:  ${url}`, LogLevel.Trace);
+
+    // await request({ type: 'PUT', url,  })
+}
+
+export async function generateCompleteManifest(output: Log, pathToTgz: string): Promise<{ manifestObj: OCIManifest; manifestStr: string; digest: string } | undefined> {
+
+    const tgzLayer = await calculateTgzLayer(output, pathToTgz);
+    if (!tgzLayer) {
+        output.write(`Failed to calculate tgz layer.`, LogLevel.Error);
+        return undefined;
+    }
+
+    return await calculateContentDigest(output, tgzLayer);
+}
+
+export async function calculateTgzLayer(output: Log, pathToTgz: string): Promise<OCILayer | undefined> {
+    output.write(`Creating manifest from ${pathToTgz}`, LogLevel.Trace);
+    if (!(await isLocalFile(pathToTgz))) {
+        output.write(`${pathToTgz} does not exist.`, LogLevel.Error);
+        return undefined;
+    }
+
+    const tarBytes = fs.readFileSync(pathToTgz);
+
+    const tarSha256 = crypto.createHash('sha256').update(tarBytes).digest('hex');
+    output.write(`${pathToTgz}:  sha256:${tarSha256} (size: ${tarBytes.byteLength})`, LogLevel.Trace);
+
+    return {
+        mediaType: 'application/vnd.devcontainers.layer.v1+tar',
+        digest: `sha256:${tarSha256}`,
+        size: tarBytes.byteLength,
+        annotations: {
+            'org.opencontainers.image.title': path.basename(pathToTgz),
+        }
+    };
+}
+
+// Spec: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#checking-if-content-exists-in-the-registry
+    //       Requires registry auth token.
+    export async function checkIfBlobExists(output: Log, featureRef: OCIFeatureRef, digest: string, authToken: string): Promise<boolean> {
+    const headers: HEADERS = {
+        'user-agent': 'devcontainer',
+        'authorization': `Bearer ${authToken}`,
+    };
+
+    const url = `https://${featureRef.registry}/v2/${featureRef.namespace}/${featureRef.featureName}/blobs/${digest}`;
+    const statusCode = await headRequest({ url, headers }, output);
+
+    output.write(`${url}: ${statusCode}`, LogLevel.Trace);
+    return statusCode === 200;
+}
+
+    // Spec: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#post-then-put
+    //       Requires registry auth token.
+    export async function postUploadSessionId(output: Log, featureRef: OCIFeatureRef, authToken: string): Promise<string | undefined> {
```

Comment:
A couple of functions in this file are indentend in a weird way, like `checkIfBlobExists`, `postUploadSessionId`.